proposals/006-filter-api-to-expose-client-sasl-info.md

```diff
@@ -0,0 +1,166 @@
+     * After calling this method the result of {@link #clientSaslContext()} will
+     * be non-empty for this and other filters.
+     * 
+     * In order to support reauthentication, calls to this method and 
+     * {@link #clientSaslAuthenticationFailure(String, String, Exception)}
+     * may be arbitrarily interleaved during the lifetime of a given filter instance.
+     *
+     * This method can only be called by filters created from {@link FilterFactory FilterFactories} which 
+     * have been annotated with {@link ClientSaslProducer @ClientSaslProducer}. 
+     * Calls from filters where this is not the case will be logged but otherwise ignored.
+     *
+     * @param mechanism The SASL mechanism used.
+     * @param authorizationId The authorization ID.
+     */
+    void clientSaslAuthenticationSuccess(String mechanism, String authorizationId);
+
+    /**
+     * Allows a filter (typically one which implements {@link SaslAuthenticateRequestFilter})
+     * to announce a failed authentication outcome with the Kafka client.
+     * It is the filter's responsilbity to return the right error response to a client, and/or disconnect.
+     * 
+     * In order to support reauthentication, calls to this method and 
+     * {@link #clientSaslAuthenticationSuccess(String, String)}
+     * may be arbitrarily interleaved during the lifetime of a given filter instance.
+     *
+     * This method can only be called by filters created from {@link FilterFactory FilterFactories} which 
+     * have been annotated with {@link ClientSaslProducer @ClientSaslProducer}. 
+     * Calls from filters where this is not the case will be logged but otherwise ignored.
+     *
+     * @param mechanism The SASL mechanism used, or null if this is not known.
+     * @param authorizationId The authorization ID, or null if this is not known.
+     * @param exception An exception describing the authentication failure.
+     */
+    void clientSaslAuthenticationFailure(String mechanism, String authorizationId, Exception exception);
+```
+
+Note that [RFC 4422][RFC4422] defines the "authorization identity" as one of pieces of information transferred via the challenges and responses defined by a mechanism. 
+In SASL this need not be the same thing as a client's username or other identity, though it usually is in Apache Kafka's use of SASL.
+We're sticking with the SASL terminology in this part of the API.
+
+### APIs for Filters to consume client SASL information
+
+Some filters, such as audit loggers, may need to use SASL authentication information specifically.
+For such filters we will add the following methods to `FilterContext`:
+
+```java
+    /**
+     * Returns the SASL context for the client connection, or empty if the client
+     * has not successfully authenticated using SASL.
+     * Filters should use {@link #clientPrincipal()} in preference to this method, unless they require SASL-specific functionality.
```

Comment:
Maybe the clientPrincipal reference should go from this doc, it's not defined here or in 004